frontend/__tests__/components/environment.spec.tsx

```diff
@@ -0,0 +1,74 @@
+import * as React from 'react';
+import { shallow, mount } from 'enzyme';
+
+import { EnvironmentPage } from '../../public/components/environment';
+
+describe('EnvironmentPage', () => {
+
+  const objects={
+    kind: '',
+    apiVersion: '',
+    metadata: {},
+    spec: {},
+    status: {}
+  };
+
+  const environmentPageRO=<EnvironmentPage
+    obj={objects}
+    rawEnvData={[ { 'env': [ { 'name': 'test', 'value': ':0', 'ID': 0 } ] } ]}
+    envPath={[]}
+    readOnly={true}
+  />;
+
+  const environmentPage=<EnvironmentPage
+    obj={objects}
+    rawEnvData={[ { 'env': [ { 'name': 'test', 'value': ':0', 'ID': 0 } ] } ]}
+    envPath={['test']}
+    readOnly={false}
+  />;
+
+  it('does not show help text when page read-only', () => {
+    const wrapper=shallow(environmentPageRO);
+    expect(wrapper.find('p').exists()).toEqual(false);
+  });
+
+  it('renders save and reload buttons when page not read-only', () => {
+    const wrapper=shallow(environmentPage);
+    expect(wrapper.find('.environment-buttons button').exists()).toEqual(true);
+  });
+
+  it('renders error message when error in state', () => {
+    const wrapper=shallow(environmentPage);
+    wrapper.setState({errorMessage: 'errorMessage'});
+    expect(wrapper.find('.environment-buttons p').text()).toContain('errorMessage');
+  });
+
+  it('renders error message when data is stale', () => {
+    const wrapper=shallow(environmentPage);
+    wrapper.setState({stale: 'stale'});
+    expect(wrapper.find('.environment-buttons p').text()).toContain('The information on this page is no longer current. Click Reload to update and lose edits, or Save Changes to overwrite.');
+  });
+
+  it('renders success message when data is updated successfully', () => {
+    const wrapper=shallow(environmentPage);
+    wrapper.setState({success: 'success'});
+    expect(wrapper.find('.environment-buttons p').text()).toContain('success');
+  });
+
+  it('renders container headlines successfully', () => {
+    const wrapper=shallow(environmentPage);
+    expect(wrapper.find('.co-m-pane__body-group h1').exists()).toBe(true);
+  });
+
+  it('gives the AsyncComponent the correct environment value editor props', () => {
+    const wrapper=mount(environmentPage);
+    expect(wrapper.find('AsyncComponent').props().nameValuePairs[0][0]).toBe('test');
+  });
+
+  it('renders PairElements successfully', () => {
+    setTimeout(() => {
```

Comment:
Why does this one need the `setTimeout()` wrapper?

Reply:
The Environment editor asynchronously imports NameValueEditor (which handles the input fields (PairElements)). Enzyme was rendering the Environment page before the import happened, so they weren't showing up without the timeout.

Reply:
Ah, ok.  So is there a way to codify this?  The `NameValueEditor` isn't referenced in this file, and `setTimeout(1000)` is a bit of a magic number... seems like a race condition. It works now, but could potentially break in the future, and it wouldn't be immediately obvious why.

Reply:
If you use an asynchronous callback in a Jasmine test, you must pass the `done` function and call it once you finish all the `expect()` calls. See [this test for `AsyncComponent`](https://github.com/openshift/console/blob/master/frontend/__tests__/components/utils/async.spec.tsx#L16) and similar tests. The better solution would be to use Jasmine "spies" to mock the async call and return a custom promise with your assertions.

Reply:
It uses AsyncComponent for the import, so do I even need to test this? As long as AsyncComponent is working correctly, it should work, right?